datafusion/physical-plan/src/aggregates/group_values/column.rs

```diff
@@ -196,6 +570,324 @@ impl GroupValues for GroupValuesColumn {
+
+        self.group_values[0].len()
+    }
+
+    fn emit(&mut self, emit_to: EmitTo) -> Result<Vec<ArrayRef>> {
+        let mut output = match emit_to {
+            EmitTo::All => {
+                let group_values = mem::take(&mut self.group_values);
+                debug_assert!(self.group_values.is_empty());
+
+                group_values
+                    .into_iter()
+                    .map(|v| v.build())
+                    .collect::<Vec<_>>()
+            }
+            EmitTo::First(n) => {
+                let output = self
+                    .group_values
+                    .iter_mut()
+                    .map(|v| v.take_n(n))
+                    .collect::<Vec<_>>();
+                let new_group_index_lists =
+                    Vec::with_capacity(self.group_index_lists.len());
+                let old_group_index_lists =
+                    mem::replace(&mut self.group_index_lists, new_group_index_lists);
+
+                // SAFETY: self.map outlives iterator and is not modified concurrently
+                unsafe {
+                    for bucket in self.map.iter() {
+                        // Check if it is `inlined` or `non-inlined`
+                        if bucket.as_ref().1.is_non_inlined() {
+                            // Non-inlined case
+                            // We take `group_index_list` from `old_group_index_lists`
+                            let list_offset = bucket.as_ref().1.value() as usize;
+                            let old_group_index_list =
+                                &old_group_index_lists[list_offset];
+
+                            let mut new_group_index_list = Vec::new();
+                            for &group_index in old_group_index_list {
+                                if let Some(remaining) = group_index.checked_sub(n) {
+                                    new_group_index_list.push(remaining);
+                                }
+                            }
+
+                            // The possible results:
+                            //   - `new_group_index_list` is empty, we should erase this bucket
+                            //   - only one value in `new_group_index_list`, switch the `view` to `inlined`
+                            //   - still multiple values in `new_group_index_list`, build and set the new `unlined view`
+                            if new_group_index_list.is_empty() {
+                                self.map.erase(bucket);
+                            } else if new_group_index_list.len() == 1 {
+                                let group_index = new_group_index_list.first().unwrap();
+                                bucket.as_mut().1 =
+                                    GroupIndexView::new_inlined(*group_index as u64);
+                            } else {
+                                let new_list_offset = self.group_index_lists.len();
+                                self.group_index_lists.push(new_group_index_list);
```

Comment:
I wonder could we replace the `old_group_index_list` with `new_group_index_list`? The old group index becomes *garbage*, if we replace it with the new group indexes we don't need garbage collection for it.

Reply:
🤔 current when taking n happen, we need to create an `new group_index_lists`,  and push the `group_index_list` still not empty from `old group_index_lists` to it.

The alternative is that, we don't create `new group_index_lists`, and just resue the `old group_index_lists`.
But I am afraid it will become something like following finally:
```text
[],
[1, 2],
[],
...
[],
[],
[3, 4],
```


I think we can actually reuse the `group_index_list` taken from `old group_index_lists`, rather than dropping it:
- We can define something like `taken_group_index_list_buffer`
- And we will place the `kept group indices` into them
- Finally we copy the `kept group indices` from taken_group_index_list_buffer` to the reused `group_index_list`,
  and push it to the current `grou_index_lists`.

Reply:
I have another idea like https://github.com/Rachelint/arrow-datafusion/pull/2/files, but given the test doesn't seem to cover the code here (I didn't see the list offset print out), so we need to strengthen test to make sure the change makes sense.